cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,725 @@
+	}
+
+	// we don't commit on route publishes, since we don't keep track of which
+	// routes we've sent to our peers in the state.
+}
+
+// Authenticate checks that the given [Auth] message is valid and proves
+// knowledge of the shared secret. If this check is passed, we allow mutation of
+// this [AssembleState] via future calls to [AssembleState.VerifyPeer] with the same
+// certificate.
+//
+// An error is returned if the message's HMAC is found to not prove knowledge of
+// the shared secret.
+func (as *AssembleState) Authenticate(auth Auth, cert []byte) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	fp := CalculateFP(cert)
+
+	expectedHMAC := CalculateHMAC(auth.RDT, fp, as.config.Secret)
+	if !hmac.Equal(expectedHMAC, auth.HMAC) {
+		return errors.New("received invalid HMAC from peer")
+	}
+
+	if _, ok := as.trusted[fp]; ok {
+		if as.trusted[fp].RDT != auth.RDT {
+			return fmt.Errorf("peer with rdt %v is using a new TLS certificate", auth.RDT)
+		}
+	} else {
+		as.trusted[fp] = peer{
+			RDT:  auth.RDT,
+			Cert: cert,
+		}
+	}
+
+	as.fingerprints[auth.RDT] = fp
+
+	// if we have discovered the route to this peer, we should record an
+	// authoritative route to it. this ensures that we send the route from our
+	// local node to this peer when we publish
+	if addr, ok := as.addresses[fp]; ok {
+		as.selector.AddAuthoritativeRoute(auth.RDT, addr)
+	}
+
+	as.commit()
+
+	as.logger.Debug("got valid auth message", "peer-rdt", auth.RDT)
+
+	return nil
+}
+
+// VerifyPeer checks if the given certificate is trusted and maps to a known RDT.
+// If it is, then a [PeerHandle] is returned that can be used to modify the state
+// of the cluster on this peer's behalf.
+//
+// An error is returned if the certificate isn't trusted.
+func (as *AssembleState) VerifyPeer(cert []byte) (*PeerHandle, error) {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	fp := CalculateFP(cert)
+
+	p, ok := as.trusted[fp]
+	if !ok {
+		return nil, errors.New("given TLS certificate is not associated with a trusted RDT")
+	}
+
+	return &PeerHandle{
+		as:   as,
+		peer: p.RDT,
+	}, nil
+}
+
+// Run starts the assembly process, managing both the server and periodic client operations.
+// It returns when the context is cancelled, returning the final routes discovered.
+func (as *AssembleState) Run(
+	ctx context.Context,
+	transport Transport,
+	discover Discoverer,
+) (Routes, error) {
+	addr := fmt.Sprintf("%s:%d", as.config.IP, as.config.Port)
+	client := transport.NewClient(as.cert)
+
+	var wg sync.WaitGroup
+
+	// start the server that handles incoming requests
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := transport.Serve(ctx, addr, as.cert, as); err != nil {
+			as.logger.Error(err.Error())
+		}
+	}()
+
+	// start periodic discovery of peers.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		periodic(ctx, time.Second*5, time.Second*1, func(ctx context.Context) {
```

Comment:
This is effectively an interval from 4 to 6 seconds right? Why do we wait for at least 4 seconds before anything happens? Or is that not right?